### ISSUE TITLE
async-io-test.c++ failure (capnproto 1.0.1, Debian Bookworm)

### ISSUE THREAD

**Issue by GitMensch** [capnproto/capnproto#1781]
~~~
[ TEST ] kj/async-io-test.c++:3069: OS handle pumpTo
[ PASS ] kj/async-io-test.c++:3069: OS handle pumpTo (1247 μs)
[ TEST ] kj/async-io-test.c++:3124: OS handle pumpTo small limit
[ PASS ] kj/async-io-test.c++:3124: OS handle pumpTo small limit (96 μs)
[ TEST ] kj/async-io-test.c++:3146: OS handle pumpTo small limit -- write first then read
[ PASS ] kj/async-io-test.c++:3146: OS handle pumpTo small limit -- write first then read (69 μs)
[ TEST ] kj/async-io-test.c++:3173: OS handle pumpTo large limit
[ PASS ] kj/async-io-test.c++:3173: OS handle pumpTo large limit (1541 μs)
[ TEST ] kj/async-io-test.c++:3195: OS handle pumpTo large limit -- write first then read
[ PASS ] kj/async-io-test.c++:3195: OS handle pumpTo large limit -- write first then read (1991 μs)
[ TEST ] kj/async-io-test.c++:3241: OS handle pumpTo write buffer is full before pump
kj/async-io-test.c++:3231: failed: expected pos < huge.size() [2000000 < 2000000]; whoa, big buffer
stack: 559e31acc66e 559e31acd94d
[ FAIL ] kj/async-io-test.c++:3241: OS handle pumpTo write buffer is full before pump (4344 μs)
[ TEST ] kj/async-io-test.c++:3275: OS handle pumpTo write buffer is full before pump -- and pump ends early
kj/async-io-test.c++:3231: failed: expected pos < huge.size() [2000000 < 2000000]; whoa, big buffer
stack: 559e31acc66e 559e31acd0e7
[ FAIL ] kj/async-io-test.c++:3275: OS handle pumpTo write buffer is full before pump -- and pump ends early (2996 μs)
[ TEST ] kj/async-io-test.c++:3304: OS handle pumpTo write buffer is full before pump -- and pump hits limit early
kj/async-io-test.c++:3231: failed: expected pos < huge.size() [2000000 < 2000000]; whoa, big buffer
stack: 559e31acc66e 559e31acc977
[ FAIL ] kj/async-io-test.c++:3304: OS handle pumpTo write buffer is full before pump -- and pump hits limit early (3425 μs)
[ TEST ] kj/async-io-test.c++:3332: OS handle pumpTo write buffer is full before pump -- and a lot of data is pumped
kj/async-io-test.c++:3231: failed: expected pos < huge.size() [2000000 < 2000000]; whoa, big buffer
stack: 559e31acc66e 559e31acf48d
[ FAIL ] kj/async-io-test.c++:3332: OS handle pumpTo write buffer is full before pump -- and a lot of data is pumped (4095 μs)
[ TEST ] kj/async-io-test.c++:3363: pump file to socket
[ PASS ] kj/async-io-test.c++:3363: pump file to socket (7477 μs)
~~~

Can be reproduced by building the tag on gitpod.io.

**Comment by kentonv** [comment]
It looks like this system's kernel is configured to give Unix sockets a default buffer size greater than 2MB. This is unusual; a more typical configuration is 64kb. There's probably some sysctl that controls this.

The test is confused because it's trying to fill up a socket buffer but it only tries to write 2,000,000 bytes. If the socket accepts all of those bytes, the test concludes that something is wrong and bails out. But probably nothing is actually wrong here, it's just that your system is actually using a giant buffer. Increasing the number on line 3227 of async-io-test.c++ might fix the test. If you want to try that, I'm happy to accept a PR.

That said, this is only a problem with the test. There's no problem in the actual Cap'n Proto implementation. So, you can safely ignore this failure.

**Comment by GitMensch** [comment]
Comparing this environment

```sh
$ ls /proc/sys/net/ipv4/tcp_*mem
/proc/sys/net/ipv4/tcp_rmem  /proc/sys/net/ipv4/tcp_wmem
$ cat /proc/sys/net/ipv4/tcp_*mem
4096    87380   6291456
4096    16384   4194304
```

to my normal dev box

```sh
$ ls /proc/sys/net/ipv4/tcp_*mem
/proc/sys/net/ipv4/tcp_mem  /proc/sys/net/ipv4/tcp_rmem  /proc/sys/net/ipv4/tcp_wmem
$ cat /proc/sys/net/ipv4/tcp_*mem
190212  253616  380424
4096    131072  6291456
4096    16384   4194304
```

shows quite identical values. Increasing the 2'000'000 to 4'200'000 leads to those tests passing.

Do you need a PR for that (I have no clue if this increase may result in a downside on other systems)?


Nonetheless the final output is:

```
1046 test(s) passed 
5 test(s) failed 
FAIL: capnp-test
```

How can I rerun those tests verbose?

**Comment by kentonv** [comment]
> `$ ls /proc/sys/net/ipv4/tcp_*mem`

The socket in question is a unix socket, not a TCP socket. So, the TCP settings are not expected to affect it. I'm not entirely sure what sysctl would affect a unix socket, the [man page](https://man7.org/linux/man-pages/man7/unix.7.html) doesn't seem to mention any, so it does seem quite odd that the behavior differs on this system.

> Do you need a PR for that (I have no clue if this increase may result in a downside on other systems)?

I think the only harm is allocating more memory for the test, but that's not a big deal. I would accept the PR.

> How can I rerun those tests verbose?

There isn't a separate verbose mode, just the log which you quoted before. Which tests are still failing?

**Comment by GitMensch** [comment]
> Which tests are still failing?

Retested on current master

~~~
[ TEST ] kj/filesystem-disk-test.c++:630: DiskDirectory copy
[ PASS ] kj/filesystem-disk-test.c++:630: DiskDirectory copy (2004 μs)
[ TEST ] kj/filesystem-disk-test.c++:654: DiskDirectory copy-replace
[ FAIL ] kj/filesystem-disk-test.c++:654: DiskDirectory copy-replace (3315 μs)
[ TEST ] kj/filesystem-disk-test.c++:691: DiskDirectory move
[ PASS ] kj/filesystem-disk-test.c++:691: DiskDirectory move (751 μs)
[ TEST ] kj/filesystem-disk-test.c++:710: DiskDirectory move-replace
[ FAIL ] kj/filesystem-disk-test.c++:710: DiskDirectory move-replace (924 μs)
[ TEST ] kj/filesystem-disk-test.c++:742: DiskDirectory createTemporary
[ PASS ] kj/filesystem-disk-test.c++:742: DiskDirectory createTemporary (269 μs)
[ TEST ] kj/filesystem-disk-test.c++:752: DiskDirectory replaceSubdir()
[ FAIL ] kj/filesystem-disk-test.c++:752: DiskDirectory replaceSubdir() (1004 μs)
[ TEST ] kj/filesystem-disk-test.c++:793: DiskDirectory replace directory with file
[ FAIL ] kj/filesystem-disk-test.c++:793: DiskDirectory replace directory with file (525 μs)
[ TEST ] kj/filesystem-disk-test.c++:821: DiskDirectory replace file with directory
[ FAIL ] kj/filesystem-disk-test.c++:821: DiskDirectory replace file with directory (1081 μs)
[ TEST ] kj/filesystem-disk-test.c++:967: DiskFilesystem::computeCurrentPath
~~~

new issue?

**Comment by kentonv** [comment]
That's very strange that they're reporting failures but not logging anything. Is it possible you have stderr redirected elsewhere? This log looks like it contains the stdout output but not the stderr output.

**Comment by GitMensch** [comment]
```
[ TEST ] kj/filesystem-disk-test.c++:280: DiskFile
[ PASS ] kj/filesystem-disk-test.c++:280: DiskFile (1515 μs)
[ TEST ] kj/filesystem-disk-test.c++:373: DiskFile::copy()
[ PASS ] kj/filesystem-disk-test.c++:373: DiskFile::copy() (974 μs)
[ TEST ] kj/filesystem-disk-test.c++:398: DiskDirectory
[ PASS ] kj/filesystem-disk-test.c++:398: DiskDirectory (1170 μs)
[ TEST ] kj/filesystem-disk-test.c++:546: DiskDirectory symlinks
[ PASS ] kj/filesystem-disk-test.c++:546: DiskDirectory symlinks (689 μs)
[ TEST ] kj/filesystem-disk-test.c++:602: DiskDirectory link
[ PASS ] kj/filesystem-disk-test.c++:602: DiskDirectory link (791 μs)
[ TEST ] kj/filesystem-disk-test.c++:630: DiskDirectory copy
[ PASS ] kj/filesystem-disk-test.c++:630: DiskDirectory copy (2122 μs)
[ TEST ] kj/filesystem-disk-test.c++:654: DiskDirectory copy-replace
kj/filesystem-disk-unix.c++:1135: failed: renameat2(fromPath, toPath, EXCHANGE): File exists; fromPath = .kj-tmp.900.6.link.partial; toPath = link
stack: 7feff8b27c1b 7feff8b2c04a 7feff8b162d3 7feff8b1680d 7feff8b2979f 7feff8b29c62 7feff8b10c43 5593714f1716
[ FAIL ] kj/filesystem-disk-test.c++:654: DiskDirectory copy-replace (2253 μs)
[ TEST ] kj/filesystem-disk-test.c++:691: DiskDirectory move
[ PASS ] kj/filesystem-disk-test.c++:691: DiskDirectory move (626 μs)
[ TEST ] kj/filesystem-disk-test.c++:710: DiskDirectory move-replace
kj/filesystem-disk-unix.c++:1487: failed: rename(fromPath, toPath): File exists; fromPath = foo; toPath = link
stack: 7feff8b299a4 7feff8b29c62 7feff8b10c43 5593714ef150
[ FAIL ] kj/filesystem-disk-test.c++:710: DiskDirectory move-replace (766 μs)
[ TEST ] kj/filesystem-disk-test.c++:742: DiskDirectory createTemporary
[ PASS ] kj/filesystem-disk-test.c++:742: DiskDirectory createTemporary (183 μs)
[ TEST ] kj/filesystem-disk-test.c++:752: DiskDirectory replaceSubdir()
kj/filesystem-disk-unix.c++:1135: failed: renameat2(fromPath, toPath, EXCHANGE): File exists; fromPath = .kj-tmp.900.10.foo.partial; toPath = foo
stack: 7feff8b27c1b 7feff8b2c04a 5593714ed219
[ FAIL ] kj/filesystem-disk-test.c++:752: DiskDirectory replaceSubdir() (783 μs)
[ TEST ] kj/filesystem-disk-test.c++:793: DiskDirectory replace directory with file
kj/filesystem-disk-unix.c++:1135: failed: renameat2(fromPath, toPath, EXCHANGE): Is a directory; fromPath = .kj-tmp.900.12.foo.partial; toPath = foo
stack: 7feff8b27c1b 7feff8b2c15a 5593714f7bee
[ FAIL ] kj/filesystem-disk-test.c++:793: DiskDirectory replace directory with file (479 μs)
[ TEST ] kj/filesystem-disk-test.c++:821: DiskDirectory replace file with directory
kj/filesystem-disk-unix.c++:1135: failed: renameat2(fromPath, toPath, EXCHANGE): Not a directory; fromPath = .kj-tmp.900.14.foo.partial; toPath = foo
stack: 7feff8b27c1b 7feff8b2c04a 5593714ebae5
[ FAIL ] kj/filesystem-disk-test.c++:821: DiskDirectory replace file with directory (616 μs)
[ TEST ] kj/filesystem-disk-test.c++:967: DiskFilesystem::computeCurrentPath
kj/filesystem-disk-unix.c++:1734: PWD environment variable doesn't match current directory; pwd = /workspace/obj
[ PASS ] kj/filesystem-disk-test.c++:967: DiskFilesystem::computeCurrentPath (486 μs)
```

**Comment by GitMensch** [comment]
that is XFS (within a docker setup)

**Comment by kentonv** [comment]
Hmm, it's possible this code doesn't work correctly on XFS, perhaps due to the use of advanced APIs that XFS implements with subtly different behavior than other filesystems. I'm not sure if I've personally tested it on XFS before; I have tested ext4, btrfs, and tmpfs.

This seems like a bug that may need some work to fix.

FWIW it probably won't affect any real use of Cap'n Proto unless your app is explicitly using the filesystem APIs.

**Comment by GitMensch** [comment]
Fixed by 2ed5cdc.
